### PR TITLE
RFC: Remove special functions from Base

### LINKED ISSUE #8536: move special functions out of Base?
Continuing the discussion in #8294, now that the [amos code has been converted to julia](http://github.com/nolta/SpecialFunctions.jl), do we want to start moving the special functions (e.g., `airy*`, `bessel*`, `hankel*`) out of Base and into a package?

### PR DESCRIPTION
**Pull request by ararslan**
This PR removes bessel, hankel, airy, error, dawson, eta, zeta, digamma and its inverse, trigamma, and polygamma functions from Base as part of #19598. The dependency on openspecfun will remain until `rem2pi` can be written without calling into openspecfun.

Fixes #8536

### CODE REVIEW

File: base/exports.jl [hunk 1]
```diff
@@ -432,37 +422,8 @@ export
...
-    besselj0,
-    besselj1,
-    besseljx,
-    besselk,
-    besselkx,
-    bessely,
-    bessely0,
-    bessely1,
-    besselyx,
     beta,
```
Comment (JeffBezanson) [inline base/exports.jl]: Should we move `beta` as well?

**Comment by JeffBezanson** [comment]
Looks like `libspecfun.a` is accidentally checked in here.

File: base/exports.jl (hunk 1 above)
Comment (ararslan) [inline base/exports.jl]: I thought about that, but those are simple extensions of `gamma`/`lgamma`, so in the interest of not doing _too_ much at once, I left those for now.

**Comment by ararslan** [comment]
Ah whoops, thanks. Will fix.

**Comment by tkelman** [comment]
Why do we need to check in openspecfun source, couldn't we leave those functions ccalls as-is until there are pure Julia replacements? I don't see much advantage in moving a subset of the C code into this repository when the build machinery for openspecfun already exists and is pretty well-proven by now.

File: src/Makefile [hunk 2]
```diff
@@ -203,12 +203,18 @@ $(BUILDDIR)/support/libsupport.a: $(SRCDIR)/support/*.h $(SRCDIR)/support/*.c
 $(BUILDDIR)/support/libsupport-debug.a: $(SRCDIR)/support/*.h $(SRCDIR)/support/*.c
 	$(MAKE) -C $(SRCDIR)/support debug BUILDDIR='$(abspath $(BUILDDIR)/support)'
 
+$(BUILDDIR)/specfun/libspecfun.a: $(SRCDIR)/specfun/*.h $(SRCDIR)/specfun/*.c $(BUILDDIR)/support/libsupport.a
```
Comment (tkelman) [inline src/Makefile]: does this actually depend on libsupport?

Comment (ararslan) [inline src/Makefile]: Nope

**Comment by ararslan** [comment]
> Why do we need to check in openspecfun source, couldn't we leave those functions ccalls as-is until there are pure Julia replacements? I don't see much advantage in moving a subset of the C code into this repository when the build machinery for openspecfun already exists and is pretty well-proven by now.

Fair points. The advantage of moving just `rem2pi` stuff into libjulia is to remove the dependency on openspecfun, since we'd then only be using it for one function, which seems silly. But I'm fine reverting all of that and just leaving openspecfun as-is if that's what's best.

**Comment by oxinabox** [comment]
> (See also simonbyrne/libm.jl for a mostly complete implementation of the requisite functionality in Julia.)

Do you mean @simonbyrne 's 2 years stale repo, or one of the more recent works? (which have also involved @simonbyrne )

 - https://github.com/musm/Amal.jl
 - https://github.com/musm/Sleef.jl
 - https://github.com/JuliaMath/Libm.jl/

**Comment by tkelman** [comment]
Moving C code around and getting the build system to be happy with it can be a fair amount of work, especially if the hopefully-not-too-distant plan is to replace it anyway? Openspecfun will be good to get rid of particularly since it's a Fortran library (so if it gets moved to a package where your plan is to compile it from source, you can expect a lot of users to not have a Fortran compiler installed), but it sounds like we're not all the way there yet.

**Comment by ararslan** [comment]
@tkelman I have indeed put quite a bit of work into getting it right, and the build system is happy with it in my local tests (current CI failure is because I forgot to remove the docs, so doctests are failing), but I understand the concern.

Yes, my plan for the SpecialFunctions package is for users to compile from source unless:

1. They're on macOS, in which case there's a Homebrew formula for openspecfun with bottles, or

2. They're on Windows, in which case I'm hoping we can cross-compile and host some binaries on BinTray.

I think gfortran is still included in most Linux distros, so I'm not too worried about Linux users being left in the cold. Even if it isn't, I would venture a guess that most Linux users could figure out how to install gfortran and compile.

@oxinabox Yes, I meant Simon's libm.jl, which includes a mostly complete implementation of the only function we need here. It's only missing Payne-Hanek reduction for large arguments. If another package has a complete implementation, that's great.

File: src/Makefile (hunk 2 above)
Comment (tkelman) [inline src/Makefile]: then the makefile rule shouldn't be written as if it does

**Comment by StefanKarpinski** [comment]
I'm also fine with continuing to build and ship `libopenspecfun` with Julia and just moving the Julia code that uses that library into a package, which would allow `rem2pi` to continue using `libopenspecfun`. That's similar to the `libRmath` situation we had for a long time. Hopefully this will take a bit less time to sort out, but there's no need to get to the perfect state here.

**Comment by ararslan** [comment]
@StefanKarpinski Yeah, Tony and I chatted about that yesterday. I think that's a good course of action, especially since SpecialFunctions.jl can then continue to use libopenspecfun without having to use BinDeps for 0.6.

**Comment by StefanKarpinski** [comment]
Great. That's a plan then.

**Comment by ararslan** [comment]
Things are good to go on this end, and the package will (hopefully) soon be registered! See https://github.com/JuliaLang/METADATA.jl/pull/7859

File: base/deprecated.jl [hunk 3]
```diff
@@ -672,65 +661,6 @@ end
 # Deprecate isimag (#19947).
 @deprecate isimag(z::Number) iszero(real(z))
 
-@deprecate airy(z::Number) airyai(z)
-@deprecate airyx(z::Number) airyaix(z)
-@deprecate airyprime(z::Number) airyaiprime(z)
```
Comment (tkelman) [inline base/deprecated.jl]: these deprecations are fairly new and should stay

File: base/deprecated.jl [hunk 4]
```diff
@@ -226,14 +226,9 @@ for f in (
         # base/special/log.jl
         :log, :log1p,
         # base/special/gamma.jl
-        :gamma, :lfact, :digamma, :trigamma, :zeta, :eta,
-        # base/special/erf.jl
-        :erfcx, :erfi, :dawson,
-        # base/special/bessel.jl
-        :airyai, :airyaiprime, :airybi, :airybiprime,
-        :besselj0, :besselj1, :bessely0, :bessely1,
```
Comment (tkelman) [inline base/deprecated.jl]: does the package defined deprecated vectorized versions of these?

Comment (ararslan) [inline base/deprecated.jl]: No, didn't seem necessary but I can add them to the package if you think that'd be best.

File: base/deprecated.jl (hunk 3 above)
Comment (ararslan) [inline base/deprecated.jl]: How should I deprecate them if the functions they should be deprecated in favor of are no longer in Base? Something like
```julia
function airy(z::Number)
    error("airy(z) has been deprecated in favor of airyai(z) in the SpecialFunctions package.")
end
```
?

Comment (ararslan) [inline base/deprecated.jl]: Oh, but I guess the deprecation targets are still exported from Base but themselves emit errors on use

File: base/deprecated.jl (hunk 4 above)
Comment (tkelman) [inline base/deprecated.jl]: we haven't had a release with them deprecated, it's only been on master, so I would do that for the sake of anyone upgrading

**Comment by ararslan** [comment]
Per a chat with Tony, all deprecations have been moved from here to the package and symbols already deprecated here are just added to the list here that emits an error pointing to SpecialFunctions. The functions are properly deprecated in SpecialFunctions.

File: base/special/gamma.jl [hunk 5]
```diff
@@ -134,347 +134,9 @@ lgamma{T<:Union{Float32,Float16}}(z::Complex{T}) = Complex{T}(lgamma(Complex{Flo
...
-
-    s = m+1
-    if real(z) <= 0 # reflection formula
-        (zeta(s, 1-z) + signflip(m, cotderiv(m,z))) * (-gamma(s))
-    else
-        signflip(m, zeta(s,z) * (-gamma(s)))
-    end
-end
-
 # TODO: better way to do this
```
Comment (tkelman) [inline base/special/gamma.jl]: doesn't look like these are needed any more

File: base/special/gamma.jl [hunk 6]
```diff
@@ -134,347 +134,9 @@ lgamma{T<:Union{Float32,Float16}}(z::Complex{T}) = Complex{T}(lgamma(Complex{Flo
...
-    end
-    t = inv(z)
-    w = t * t # 1/z^2
-    ψ += t + 0.5*w
-    # the coefficients here are Float64(bernoulli[2:9])
-    ψ += t*w * @evalpoly(w,0.16666666666666666,-0.03333333333333333,0.023809523809523808,-0.03333333333333333,0.07575757575757576,-0.2531135531135531,1.1666666666666667,-7.092156862745098)
-end
-
 signflip(m::Number, z) = (-1+0im)^m * z
 signflip(m::Integer, z) = iseven(m) ? z : -z
```
Comment (ararslan) [inline base/special/gamma.jl]: Yep, you're right (here and above). Will remove. Thanks!

**Comment by tkelman** [comment]
I might miss having `erf` and `erfc` without an import a little bit, but LGTM aside from the dead code noted above, and the post-merge TODO (https://github.com/JuliaLang/METADATA.jl/pull/7859#issuecomment-278743518) on the package importing for Julia versions after this merges (and up to the deprecations getting deleted, so should also verify `isdefined` before extending anything from Base).

We should move the PR by @oxinabox over to the package as well.

File: base/special/gamma.jl (hunk 6 above)
Comment (tkelman) [inline base/special/gamma.jl]: oh but since you were importing the Base version in the package, you'll want to add these there instead

Comment (ararslan) [inline base/special/gamma.jl]: Right, that's what I figured I'd do.

**Comment by ararslan** [comment]
32-bit AppVeyor failure seems unrelated (an error in `pkg` tests). Log backed up [here](https://gist.github.com/ararslan/48fdfd622524785eb86247e3e1c06aa0). Worth restarting AV?

**Comment by StefanKarpinski** [comment]
I don't think so, that package cloning problem happens occasionally due to bad connectivity from AppVeyor machines. The failure is after all the rest of the tests passed and the package tests passed on the other Windows build, so should be fine. As long as the Travis builds are good, you're in the clear.

**Comment by stevengj** [comment]
@ararslan, could I have a collaborator invite in SpecialFunctions?

**Comment by simonbyrne** [comment]
@stevengj I sent you an invite

**Comment by ararslan** [comment]
@stevengj You've had a pending JuliaMath membership invitation for months now. 😉

**Comment by ararslan** [comment]
I'd like to get this in before something else creates a conflict and I have to rebase, causing the PR to languish in the Travis OS X queue for an eternity, so speak now or forever hold your peace. 😄

**Pull request merged**